### PR TITLE
[Parser] typed errors overhaul

### PR DESCRIPTION
**Pull request by cindyxkuang**
Fixes # .

### Motivation
continuing from this PR: [https://github.com/coinbase/rosetta-sdk-go/pull/102](https://github.com/coinbase/rosetta-sdk-go/pull/102)
and this PR: [https://github.com/coinbase/rosetta-sdk-go/pull/129](https://github.com/coinbase/rosetta-sdk-go/pull/129)

we want to use named errors in the SDK to improve the debugging experience

this PR also includes some minor changes in /keys

### Solution
added errors.go and an ErrParser() function to check if an error was thrown by the parser

### CODE REVIEW

File: parser/errors.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	ErrCoinActionMatchCoinChangeIsNil      = errors.New("coin change is nil")
+	ErrCoinActionMatchUnexpectedCoinAction = errors.New("unexpected coin action")
+
+	ErrEqualAmountsNoOperations = errors.New("cannot check equality of 0 operations")
+	ErrEqualAmountsNotEqual     = errors.New("amounts are not equal")
+
+	ErrOppositeAmountsSameSign       = errors.New("operations have the same sign")
+	ErrOppositeAmountsAbsValMismatch = errors.New("operation absolute values are not equal")
+
+	ErrEqualAddressesTooManyOperations = errors.New("cannot check equality of <= 1 operations")
```
Comment (qiwu7) [inline parser/errors.go:71]:
```suggestion
	ErrEqualAddressesTooLittleOperations = errors.New("cannot check equality of <= 1 operations")
```

Comment (cindyxkuang) [reply parser/errors.go:71]: 😅

File: parser/errors.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	for key, val := range parserErrs {
+		if findError(val, err) {
+			return true, key
+		}
+	}
+	return false, ""
+}
+
+// findError
+func findError(errorList []error, err error) bool {
```
Comment (patrick-ogrady) [inline parser/errors.go:130]: Can we make this function a util so you can use it across all the packages you are overhauling?

File: keys/errors.go [hunk 3]
```diff
@@ -37,3 +37,25 @@ var (
...
+		ErrPrivKeyLengthInvalid,
+		ErrSignUnsupportedPayloadSignatureType,
+		ErrSignUnsupportedSignatureType,
+		ErrSignFailed,
+		ErrVerifyUnsupportedPayloadSignatureType,
+		ErrVerifyUnsupportedSignatureType,
+		ErrVerifyFailed,
+	}
+
+	for _, k := range keyErrors {
```
Comment (patrick-ogrady) [inline keys/errors.go:55]: nit: use `findErr` util here

File: parser/errors.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+// whether or not the error is one thrown by the asserter
+// along with the specific source of the error
+func ErrParser(err error) (bool, string) {
+	parserErrs := map[string][]error{
+		"intent error":           IntentErrs,
+		"match operations error": MatchOpsErrs,
+	}
+
+	for key, val := range parserErrs {
+		if utils.FindError(val, err) {
```
Comment (qiwu7) [inline parser/errors.go:126]: Can we add some unit tests for this? Although this is indirectly tested here with the parser/errors_tests :)

File: parser/errors.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+	ErrExpectedOperationAccountMismatch = errors.New(
+		"intended account did not match observed account",
+	)
+	ErrExpectedOperationAmountMismatch = errors.New(
+		"intended amount did not match observed amount",
+	)
+	ErrExpectedOperationTypeMismatch = errors.New(
+		"intended type did not match observed type",
+	)
+	ErrExpectedOperationsUnexpectedOperation = errors.New("found extra operation")
```
Comment (qiwu7) [inline parser/errors.go:37]: how about rename to the following? the var name looks awkward 
```suggestion
	ErrExpectedOperationsExtraOperation = errors.New("found extra operation")
```

File: parser/errors.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+	/////////////////////////////
+	/* MATCH OPERATIONS ERRORS */
+	/////////////////////////////
+	ErrAccountMatchAccountMissing           = errors.New("account is missing")
+	ErrAccountMatchSubAccountMissing        = errors.New("SubAccountIdentifier.Address is missing")
+	ErrAccountMatchSubAccountPopulated      = errors.New("SubAccount is populated")
+	ErrAccountMatchUnexpectedSubAccountAddr = errors.New("unexpected SubAccountIdentifier.Address")
+
+	ErrMetadataMatchKeyNotFound       = errors.New("key is not present in metadata")
+	ErrMetadataMatchValueTypeMismatch = errors.New("unexpected value associated with key")
```
Comment (qiwu7) [inline parser/errors.go:59]: nit: the name and the error message doesn't seem to match 🤔

File: parser/intent.go [hunk 7]
```diff
@@ -118,7 +122,7 @@ func (p *Parser) ExpectedOperations(
 
 	if len(missingIntent) > 0 {
 		errString := fmt.Sprintf(
-			"could intent match for %v",
+			"could not intent match %v",
```
Comment (qiwu7) [inline parser/intent.go:125]:
```suggestion
			"could not intent match for %v",
```

File: parser/errors.go (hunk 6 above)
Comment (cindyxkuang) [reply parser/errors.go]: changed to `ErrMetadataMatchKeyValueMismatch`

**Review by patrick-ogrady — APPROVED** [review]
LGTM!

**Pull request merged**